MINOR: Remove types from caching stores

mjsax opened the pull request:
Removes all type information (including Serdes) from caching stores and move this "one level up" (ie, into metered stores).

Impact:
 - caching stores returns `<byte[],byte[]>` on flush
 - flush callback must deserialize the data before forwarded to downstream operator

`TupleForwarded` registered "typed" `FlushListener` on "metered stores" now (not on the caching store), and "metered stores" registers a `byte[]`-`FlushListener` on the caching store. When caching store evicts and call the listener, the "metered stores" deserializes the bytes, and fires the original "typed" `FlushListener`.

This will unblock KIP-258 PR #6152 (ie, the open discussion about flushing...)
 
Call for review @guozhangwang @bbejeck @vvcephei @ableegoldman 


### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

mjsax committed:
MINOR: remove types from caching stores

guozhangwang reviewed (commented):
I understand that the serdes used in the cached store is only for deserializations for flushing to listeners, but after trying out several ways while reviewing this PR I felt there is no elegant manner to get the deserialization elsewhere so that we can trim it from the caching store itself (I've tried TupleForwarder, ForwardCacheFlushListener).

Compared with the current proposal, I'd prefer still keeping the logic scattered in the caching stores itself.

mjsax commented:
@guozhangwang I found a better way to remove the types from caching stores and refactored this PR accordingly. Let me know what you think about this proposal.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/CachedStateStore.java:
```diff
@@ -23,6 +23,6 @@
      */
-    void setFlushListener(final CacheFlushListener<K, V> listener,
-                          final boolean sendOldValues);
+    boolean setFlushListener(final CacheFlushListener<K, V> listener,
```
We return `true` if the listener was registered, ie, if the "metered stores" wraps a "caching store".

vvcephei replied:
Clever! I tried to do this change a little while ago and got hung up on this point.

mjsax replied:
Thanks. _blushing_

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java:
```diff
@@ -16,43 +16,35 @@
-class CachingKeyValueStore<K, V> extends WrappedStateStore<KeyValueStore<Bytes, byte[]>> implements KeyValueStore<Bytes, byte[]>, CachedStateStore<K, V> {
+class CachingKeyValueStore
+    extends WrappedStateStore<KeyValueStore<Bytes, byte[]>>
+    implements KeyValueStore<Bytes, byte[]>, CachedStateStore<byte[], byte[]> {
```
Using `CachedStateStore<byte[], byte[]>` instead of `CachedStateStore< Bytes, byte[]>` because it seems cleaner overall (similar on other caching stores).

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java:
```diff
@@ -113,37 +99,92 @@ private void putAndMaybeForward(final ThreadCache.DirtyEntry entry,
     }
 
     @Override
-    public void flush() {
```
Below is just some method reordering within the class -- no actual code change.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/CachingSessionStore.java:
```diff
@@ -79,6 +65,69 @@ private void initInternal(final InternalProcessorContext context) {
         });
     }
 
+    private void putAndMaybeForward(final ThreadCache.DirtyEntry entry, final InternalProcessorContext context) {
```
move this method only.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/CachingSessionStore.java:
```diff
@@ -79,6 +65,69 @@ private void initInternal(final InternalProcessorContext context) {
+        return true;
+    }
+
+    @Override
```
Below is just some method reordering within the class -- no actual code change.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/CachingWindowStore.java:
```diff
@@ -123,24 +110,13 @@ private void putAndMaybeForward(final ThreadCache.DirtyEntry entry,
         this.sendOldValues = sendOldValues;
-    }
 
-    @Override
```
Below is just some method reordering within the class -- no actual code change.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueStore.java:
```diff
@@ -115,15 +117,22 @@ public void init(final ProcessorContext context,
 
+    @SuppressWarnings("unchecked")
     @Override
-    public void close() {
```
Only moved `close()` and `approximateNumEntries`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueStore.java:
```diff
@@ -115,15 +117,22 @@ public void init(final ProcessorContext context,
+    public boolean setFlushListener(final CacheFlushListener<K, V> listener,
+                                    final boolean sendOldValues) {
+        final KeyValueStore<Bytes, byte[]> wrapped = wrapped();
+        if (wrapped instanceof CachedStateStore) {
```
We know the wrapping order, thus, this is safe. Intercepting the callback below to deserialize.

guozhangwang replied:
Meta: I like this approach slightly more than the previous one, and wondering if we can sort-of combine these two to make the logic more readable (and also the metered store lost the topic name info when calling serde which may not be acceptable):

1) Metered store do not need to inherited CachedStateStore, instead inherit a SerdeStateStore interface which just expose the getter of the serdes (note we cannot use `serdes` since it is only constructed at initialization, we have to use `keySerde` and `valueSerde` instead`).
2) In TupleForwarder we get both the SerdeStateStore, and check if its wrapped store is a CachedStateStore, if yes call `cachedStore.setFlushListener(... listener.apply(serdeStore.keySerde().deser ...))` -- i.e. basically the code below.

WDYT?

mjsax replied:
> and also the metered store lost the topic name info when calling serde which may not be acceptable

Fixed this part. Compare my other comments.

About your proposal: that is certainly possible, it's similar (or even the same?) to the original PR introducing `TypedStore` (your suggested `SerdeStateStore` is the same thing). So I am wondering what the difference to the original proposal would be -- seems you suggest to go back to the original one.

vvcephei replied:
At a high level, our store ecosystem looks like an onion. On the outside, we have a <K,V> store, and on the inside, we have a <bytes,bytes> store. All the layers in between have different responsibilities, like changelogging, caching, add metrics, etc.

The nice thing about this PR is that it gives us one clean layer that's responsible for the transition `<K,V> <=> <bytes, bytes>`. When we need to look at the de/serialization into/out-of the stores, we have exactly one place to look.

The prior code did mostly this, but to accommodate cache flushing in conjunction with the fact that the cache layer is below the transition from objects to bytes, we had to poke a hole in the onion and tell the caching layer (a bytes layer) how to deserialize. So, then there were two layers that independently know how to de/serialize, and the onion had a hole in it.

This idea to move the serialization out to the TupleForwarder is basically the same, but in the opposite direction. Again, there are two components that need to perform serialization (the serialization layer and the tuple forwarder), and again, we need to poke a hole in the onion so that the tuple forwarder can communicate directly with an inner layer.

It's not always practical to go for a "pure" design, but if readability is the goal, then it seems like we should try to avoid mixing layers, unwrapping layers, etc. as much as possible. To be fair, this is just my take on the situation.

mjsax replied:
I also think that using two nested `CacheFlushListener` is better instead of "drilling" whole into the APIs. So I prefer to stay with the current design.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredSessionStore.java:
```diff
@@ -96,86 +99,55 @@ public void init(final ProcessorContext context,
 
+    @SuppressWarnings("unchecked")
     @Override
-    public void close() {
```
Only moved `close()` and both `findSessions()` methods

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredSessionStore.java:
```diff
@@ -96,86 +99,55 @@ public void init(final ProcessorContext context,
     }
 
     @Override
-    public void remove(final Windowed<K> sessionKey) {
```
Below is just some method reordering within the class -- no actual code change.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredWindowStore.java:
```diff
@@ -96,10 +101,22 @@ public void init(final ProcessorContext context,
 
+    @SuppressWarnings("unchecked")
     @Override
-    public void close() {
```
Only moved `close()`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredWindowStore.java:
```diff
@@ -123,10 +140,6 @@ public void put(final K key,
         }
     }
 
-    private Bytes keyBytes(final K key) {
```
Below is just some method reordering within the class -- no actual code change.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KTableAggregateTest.java:
```diff
@@ -76,14 +70,17 @@ public void testAggBasic() {
 
-
         final KTable<String, String> table1 = builder.table(topic1, consumed);
-        final KTable<String, String> table2 = table1.groupBy(MockMapper.<String, String>noOpKeyValueMapper(),
```
Only Java8 cleanup and code reformatting in this class

mjsax commented on streams/src/test/java/org/apache/kafka/streams/state/internals/CachingSessionStoreTest.java:
```diff
@@ -234,69 +238,69 @@ public void shouldForwardChangedValuesDuringFlush() {
         final Windowed<Bytes> b = new Windowed<>(keyA, new SessionWindow(1, 2));
         final Windowed<String> aDeserialized = new Windowed<>("a", new SessionWindow(2, 4));
         final Windowed<String> bDeserialized = new Windowed<>("a", new SessionWindow(1, 2));
-        final List<KeyValue<Windowed<String>, Change<String>>> flushed = new ArrayList<>();
```
replace `flushed` to reuse existing `CachingKeyValueStoreTest.CacheFlushListenerStub`

guozhangwang replied:
Nice cleanup!

mjsax commented on streams/src/test/java/org/apache/kafka/streams/state/internals/CachingSessionStoreTest.java:
```diff
@@ -234,69 +238,69 @@ public void shouldForwardChangedValuesDuringFlush() {
     public void shouldNotForwardChangedValuesDuringFlushWhenSendOldValuesDisabled() {
         final Windowed<Bytes> a = new Windowed<>(keyA, new SessionWindow(0, 0));
         final Windowed<String> aDeserialized = new Windowed<>("a", new SessionWindow(0, 0));
-        final List<KeyValue<Windowed<String>, Change<String>>> flushed = new ArrayList<>();
```
as above

mjsax commented on streams/src/test/java/org/apache/kafka/test/GenericInMemoryKeyValueStore.java:
```diff
@@ -16,24 +16,28 @@
  */
-public class GenericInMemoryKeyValueStore<K extends Comparable, V> implements KeyValueStore<K, V> {
+public class GenericInMemoryKeyValueStore<K extends Comparable, V>
+    implements KeyValueStore<K, V>, CachedStateStore<K, V> {
```
Because we assume that all "metered stores" implement `CachedStateStore` we need to add it here, too.

vvcephei replied:
Sorry, I couldn't follow... How is this a "metered store"?

Was it necessary to add the interface to get some tests to work?

mjsax replied:
It's not a metered store, but we assume a certain wrapping hierarchy and `GenericInMemoryKeyValueStore` is in the same hierarchy level as "metered stores" -- for this test case, there is no metered stores, but `GenericInMemoryKeyValueStore` takes the metered stores place in the hierarchy. So, yes, it required to get the test work.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java:
```diff
@@ -113,37 +99,92 @@ private void putAndMaybeForward(final ThreadCache.DirtyEntry entry,
     @Override
-    public void flush() {
+    public void put(final Bytes key,
+                    final byte[] value) {
```
Just to confirm: this line and below do not have any logical changes just due to git diff messing the line numbers right?

mjsax replied:
Yes.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredSessionStore.java:
```diff
@@ -96,86 +99,55 @@ public void init(final ProcessorContext context,
+        if (wrapped instanceof CachedStateStore) {
+            return ((CachedStateStore<byte[], byte[]>) wrapped).setFlushListener(
+                (key, newValue, oldValue, timestamp) -> listener.apply(
+                    SessionKeySchema.from(key, serdes.keyDeserializer(), null),
```
In caching store we are passing `topic` but here `null`, is that okay?

mjsax replied:
This slipped. We can get the topic from `serdes.topic()`. Will push a fix.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredWindowStore.java:
```diff
@@ -96,10 +101,22 @@ public void init(final ProcessorContext context,
+        if (wrapped instanceof CachedStateStore) {
+            return ((CachedStateStore<byte[], byte[]>) wrapped).setFlushListener(
+                (key, newValue, oldValue, timestamp) -> listener.apply(
+                    WindowKeySchema.fromStoreKey(key, windowSizeMs, serdes.keyDeserializer(), null),
```
Ditto here.

mjsax replied:
As above

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/TupleForwarder.java:
```diff
@@ -30,45 +30,32 @@
+            cachedStateStore = (CachedStateStore) store;
+        } else {
+            // metered stores might be wrapped with `decorator` (cf. ProcessorContextImpl)
+            cachedStateStore = (CachedStateStore) ((WrappedStateStore) store).wrapped();
```
Would it makes sense to relocate the problem of maybe-unwrapping to get a CachedStateStore into a static method in WrappedStateStore, like we did for the timestamped bytes stores ( `https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/state/internals/WrappedStateStore.java#L28` )?

vvcephei replied:
An alternative is to also make the decorators implement `CachedStateStore` and just pass through to the store they wrap. Then, we don't need to have branches like this.

In fact, we could just add `implements CachedStateStore` to `WrappedStateStore` with a default, pass-though implementation.

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/TupleForwarder.java:
```diff
@@ -30,45 +30,32 @@
  */
 class TupleForwarder<K, V> {
-    private final CachedStateStore cachedStateStore;
+    private final boolean cachingEnable;
```
nit: `cachingEnable` -> `cachingEnabled`

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/TupleForwarder.java:
```diff
@@ -30,45 +30,32 @@
                              final V oldValue) {
-        if (cachedStateStore == null) {
-            context.forward(key, new Change<>(newValue, oldValue));
+        if (cachingEnable) {
```
nit: I'm wondering if just using would suffice (IMHO slightly easier to immediately grok the meaning).
```java
if (!cachingEnable) {
     context.forward(key, new Change<>(newValue, oldValue));
 }
```

mjsax replied:
That would work, too. IMHO, it's easier to avoid negation. It's easier to read the code that way (personal taste). I also tried to use
```
if(cachingDisabled) {
   context.forward(key, new Change<>(newValue, oldValue));
}
```

but `cachingDisabled` is also weird. If you insist, I can still update it.

bbejeck replied:
No, it's fine as is, my comment is subjective.

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/TupleForwarder.java:
```diff
@@ -30,45 +30,32 @@
-        } else if (store instanceof WrappedStateStore) {
-            StateStore wrapped = ((WrappedStateStore) store).wrapped();
-
-            while (wrapped instanceof WrappedStateStore && !(wrapped instanceof CachedStateStore)) {
```
The PR replaces a `while` loop with a single statement, does this mean Streams is guaranteed to only ever have one level of a `WrappedStateStore`?

mjsax replied:
`TupleForwarder` is part of the DSL and thus we know the wrapping hierarchy, and hence, this should be save. We know that the caching store is either inside metered stores, or there is no caching store.

bbejeck reviewed (commented):
Thanks, @mjsax I've made an initial pass and have a few minor comments. 
EDIT: I meant to add I like the new approach and consider it an improvement.

I also have some follow-up questions.

Since we've pushed the deserialization up a level, unless I'm missing something, it seems we are now bound always to wrap `CachedStateStore` instances in a metered store.  If that's not the case how will we deserialize from non-metered-wrapped caching stores?

Also, since this change is limited caching stores this PR will not affect streams if caching is disabled correct?

vvcephei reviewed (commented):
This looks awesome @mjsax !

I left some thoughts about the flush listener strategy. I'm in favor of what you have presented here.

Thanks,
-John

mjsax commented:
> Since we've pushed the deserialization up a level, unless I'm missing something, it seems we are now bound always to wrap `CachedStateStore` instances in a metered store. If that's not the case how will we deserialize from non-metered-wrapped caching stores?

Yes, but this is no limitation IMHO. Note, that `CachedStateStore` is internal and developers user them only via `Stores` factory. Thus, we _know_ that a `CachedStateStore` is always wrapped with a metered store.

> Also, since this change is limited caching stores this PR will not affect streams if caching is disabled correct?

Correct. That is the whole idea of our store hierarchy.

\cc @bbejeck

mjsax committed:
Github comments and rebased

mjsax commented:
Updated this and address review comments. Also rebased to resolve conflicts. Added some new tests. Call for review again.

Based on the feedback, it seems better to use the "nested listener" approach to keep a clear separation of concerns instead of "drilling" wholes in the store hierarchy layers.

guozhangwang reviewed (approved):
Sounds good to me. Please feel free to merge after jenkins passed :)

bbejeck reviewed (approved):
Thanks @mjsax LGMT

vvcephei reviewed (approved):
LGTM!

Thanks @mjsax !
